### PR TITLE
Simplify the way ga4 tracking is added to accordions

### PR DESCRIPTION
**Pull request by AshGDS**
Hi @andysellick / @JamesCGDS 

Would you be able to review this? Thanks 👍 

## What
Previously we were passing a GTM object to each accordion's render statement inside of a frontend app's template. See an example here:
 https://github.com/alphagov/collections/blob/2304bf53592f39e35270413fdb27c00c7cb3190a/app/views/world_wide_taxons/accordion.html.erb#L31 

Instead, it's simpler to just pass a boolean - `ga4_tracking: true` - in the accordion render of these frontend apps, and then we handle the GTM object in one place inside of `govuk_publishing_components` instead.

We were also using the accordion's class to check wether we tracked `aria-expanded`. We are now using a data attribute for this, called `data-ga4-expandable`

This isn't a breaking change as the old code is still in the repo, as that code was used to add some new UA tracking so we have to leave it in.

## Why

- This gives us the benefit of only having to maintain our GTM object in one repo, and we won't have to modify templates in multiple frontend apps if our GTM schema changes.
- Reduces technical debt
- Now only uses `data-ga4-` attributes which is better.

## Visual Changes

None.

### CODE REVIEW

File: docs/analytics-ga4/ga4-event-tracker.md
Comment (andysellick) [inline docs/analytics-ga4/ga4-event-tracker.md:70]: Struggling a bit this afternoon post-lunch so I could be completely wrong about this, but checking in the component guide I can't see any attribute on the show/hide all link anymore. Has something broken or am I missing something obvious?

File: app/views/govuk_publishing_components/components/docs/accordion.yml
Comment (andysellick) [inline app/views/govuk_publishing_components/components/docs/accordion.yml:208]: Nitpick: not sure you need this much data to demonstrate this example, a couple of steps would be fine.

File: app/views/govuk_publishing_components/components/_accordion.html.erb
Comment (andysellick) [inline app/views/govuk_publishing_components/components/_accordion.html.erb:28]: I think there should be a test in the `.spec` file to check that these attributes are added correctly, along with the individual attributes added for each step further down.

File: app/views/govuk_publishing_components/components/_accordion.html.erb
Comment (andysellick) [inline app/views/govuk_publishing_components/components/_accordion.html.erb:51]: Does this need a bit more in the JS test to check that these attributes get copied onto the show/hide all button correctly? (I haven't checked, there might already be something from the previous iteration, don't know if it would need updating or not).

**Review by andysellick — COMMENTED** [review]
My brain's gone a bit foggy this afternoon so this might not be particularly thorough or coherent, but looks good, like the approach, this will definitely save us some effort - just a few comments.

File: docs/analytics-ga4/ga4-event-tracker.md
Comment (AshGDS) [reply docs/analytics-ga4/ga4-event-tracker.md:70]: Good spot - I am getting confused as well. It looks like the initial implementation of `show_all_attributes` takes a JSON containing JSONs, and this would apply each JSON as a `data-` attribute to the Show All link. However we just need to apply the one JSON directly on the data attribute now.

I've also just done a git blame and realised the `data-show-all-attributes` isn't needed for UA - you introduced it specifically for GA4. I got confused because the example wasn't in our GTM schema, and I think that's just because you wrote the code for this before we had a schema, so I thought the examples were UA stuff. (See [here](https://github.com/alphagov/govuk_publishing_components/commit/f193ef04feec77e79ca666d474e0221c6fd03ba9#diff-0353e1304e7acef271c38d2ac799bf23412749d788d92d9e3e8fab8b6b02a2bbR197))

Therefore I should be able to rename the attribute to `data-ga4-show-all-attributes`

File: docs/analytics-ga4/ga4-event-tracker.md
Comment (JamesCGDS) [inline docs/analytics-ga4/ga4-event-tracker.md:73]: Minor typo - "as the accordion will contain a `data-ga4-expandable` value"

Nitpick - would it be possible to say "This sets the `action` property of the `event_data` object accordingly."? I think this comes down to personal preference though so happy to leave it as it is if you prefer.

File: docs/analytics-ga4/ga4-event-tracker.md
Comment (AshGDS) [reply docs/analytics-ga4/ga4-event-tracker.md:70]: I'm getting confused about wether I can rename `data_show_all_attributes` in the accordions, but I think I figured it out.

Initially I thought `data_show_all_attributes`  was being used for UA, so I kept it in place. Then I saw that you introduced `data_show_all_attributes` specifically for GA4, so then I thought I could remove it. But it looks like another team have used this GA4 `data_show_all_attributes` attribute to add tracking to UA, so now I can't remove it.

@andysellick the code to add attributes to the Show All attributes link takes a JSON containing JSONs, so I'll have to update the template so I can either:
- Update the template so that it passes a JSON containing a JSON to `data-show-all-attributes` (this feels like adding technical debt)
- Duplicate the JavaScript used for`data-show-all-attributes` code, and make a specific one for GA4 with the attribute name `data-ga4-show-all-attributes` (I think this is preferable)

File: docs/analytics-ga4/ga4-event-tracker.md
Comment (AshGDS) [reply docs/analytics-ga4/ga4-event-tracker.md]: Thanks @JamesCGDS, I've updated the docs

File: docs/analytics-ga4/ga4-event-tracker.md
Comment (AshGDS) [reply docs/analytics-ga4/ga4-event-tracker.md]: @andysellick I think i've resolved this, see the latest commits...

File: app/views/govuk_publishing_components/components/docs/accordion.yml
Comment (AshGDS) [reply app/views/govuk_publishing_components/components/docs/accordion.yml]: Thanks @andysellick - good point. I've made it 2 steps.

File: app/views/govuk_publishing_components/components/_accordion.html.erb
Comment (AshGDS) [reply app/views/govuk_publishing_components/components/_accordion.html.erb:28]: Thanks @andysellick - there's now a spec test for the module, `data-ga4-expandable` and `data-ga4-show-all-attributes` and the `data-ga4-event` on the 1 accordion item in the test.

**Comment by AshGDS** [comment]
Thanks @andysellick @JamesCGDS - I think I've resolved the changes you requested, feel free to re-review when you have a chance. Will squash the commits once everything is approved :+1:

File: app/assets/javascripts/govuk_publishing_components/components/accordion.js
Comment (andysellick) [inline app/assets/javascripts/govuk_publishing_components/components/accordion.js:63]: Instead of passing these attributes to the component and then moving them onto the showall button, couldn't we just hardcode them? Something like this:

```Javascript
// look for data attributes to put onto the 'show/hide all' link
var showAll = this.$module.querySelector(this.showAllControls)
var showAllAttributes = this.$module.getAttribute('data-show-all-attributes')
if (showAllAttributes) {
  try {
    var values = JSON.parse(showAllAttributes)
    var keys = Object.keys(values)
    for (var i = 0; i < keys.length; i++) {
      showAll.setAttribute('data-' + keys[i], values[keys[i]])
    }
  } catch (e) {
    console.error('Could not read accordion data attributes error: ' + e.message, window.location)
  }
}

// if GA4 is enabled, set attributes on 'show all sections' for tracking using ga4-event-tracker
var isGa4Enabled = this.$module.hasAttribute('data-ga4-expandable')
if (isGa4Enabled) {
  var indexTotal = this.$module.querySelectorAll('.govuk-accordion__section').length
  var showAllAttributesGa4 = {"event_name":"select_content","type":"accordion","index":0,"index_total": indexTotal}
  showAll.setAttribute('data-ga4-event', JSON.stringify(showAllAttributesGa4))
}
```

Comment (AshGDS) [reply app/assets/javascripts/govuk_publishing_components/components/accordion.js]: Thanks @andysellick - I've added this change and then will move it to `ga4-core.js` once the link tracker PR is merged.

File: app/assets/javascripts/govuk_publishing_components/components/accordion.js
Comment (JamesCGDS) [inline app/assets/javascripts/govuk_publishing_components/components/accordion.js:56]: This code isn't part of the PR but I just noticed it and wanted to check - is this meant to be here?

Comment (andysellick) [reply app/assets/javascripts/govuk_publishing_components/components/accordion.js:56]: Yes, it is. In the event that incorrect data is passed to the component we want an error to happen in the console.

Ideally it would be noticed by the developer implementing it, but we also have stuff to monitor this that should help us identify configuration problems like this.

**Review by JamesCGDS — APPROVED** [review]
Looks good to me 👍

File: app/views/govuk_publishing_components/components/docs/accordion.yml
Comment (andysellick) [inline app/views/govuk_publishing_components/components/docs/accordion.yml:204]: I think this now only applies (or is needed) for UA instead of GA4? Is it worth clarifying this here?

**Review by andysellick — COMMENTED** [review]
Nice work. Just two minor documentation comments then I think we're good to go 👍

File: app/assets/javascripts/govuk_publishing_components/components/accordion.js
Comment (andysellick) [reply app/assets/javascripts/govuk_publishing_components/components/accordion.js]: Does this still need moving to `ga4-core.js`? I can't think what else will need this right now. Maybe move it later when we need it, shouldn't have to block this PR.

Comment (AshGDS) [reply app/assets/javascripts/govuk_publishing_components/components/accordion.js]: This is due to `ga4-core` not being initialised in `govuk_publishing_components`, so the accordion would crash on the developer frontend which showcases the component. Therefore we decided to share the function at a later date.

**Pull request merged**
